core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java

```diff
@@ -0,0 +1,115 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.encryption;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import org.apache.iceberg.io.PositionOutputStream;
+
+public class AesGcmOutputStream extends PositionOutputStream {
+
+  private final Ciphers.AesGcmEncryptor gcmEncryptor;
+  private final PositionOutputStream targetStream;
+  private final byte[] plainBlockBuffer;
+  private final byte[] fileAadPrefix;
+
+  private int positionInBuffer;
+  private long streamPosition;
+  private int currentBlockIndex;
+
+  AesGcmOutputStream(PositionOutputStream targetStream, byte[] aesKey, byte[] fileAadPrefix)
+      throws IOException {
+    this.targetStream = targetStream;
+    this.gcmEncryptor = new Ciphers.AesGcmEncryptor(aesKey);
+    this.plainBlockBuffer = new byte[Ciphers.PLAIN_BLOCK_SIZE];
+    this.positionInBuffer = 0;
+    this.streamPosition = 0;
+    this.currentBlockIndex = 0;
+    this.fileAadPrefix = fileAadPrefix;
+
+    byte[] headerBytes =
+        ByteBuffer.allocate(Ciphers.GCM_STREAM_HEADER_LENGTH)
+            .order(ByteOrder.LITTLE_ENDIAN)
+            .put(Ciphers.GCM_STREAM_MAGIC_ARRAY)
+            .putInt(Ciphers.PLAIN_BLOCK_SIZE)
+            .array();
+    targetStream.write(headerBytes);
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    if (b.length - off < len) {
+      throw new IOException(
+          "Insufficient bytes in buffer: " + b.length + " - " + off + " < " + len);
+    }
+    int remaining = len;
+    int offset = off;
+
+    while (remaining > 0) {
+      int freeBlockBytes = Ciphers.PLAIN_BLOCK_SIZE - positionInBuffer;
+      int toWrite = freeBlockBytes <= remaining ? freeBlockBytes : remaining;
+
+      System.arraycopy(b, offset, plainBlockBuffer, positionInBuffer, toWrite);
+      positionInBuffer += toWrite;
+      if (positionInBuffer == Ciphers.PLAIN_BLOCK_SIZE) {
+        encryptAndWriteBlock();
+        positionInBuffer = 0;
+      }
+      offset += toWrite;
+      remaining -= toWrite;
+    }
+
+    streamPosition += len;
+  }
+
+  @Override
+  public long getPos() throws IOException {
+    return streamPosition;
+  }
+
+  @Override
+  public void flush() throws IOException {
+    targetStream.flush();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (positionInBuffer > 0) {
+      encryptAndWriteBlock();
+    }
+    targetStream.close();
+  }
+
+  private void encryptAndWriteBlock() throws IOException {
```

Comment:
Do you think it would be good to have a boolean `fullBlock` to keep track of whether the last block was a full block? Then we could check it here and refuse to write another block if the last block was not full.

Reply:
this function is called only for full blocks, after an explicit check:

```java
if (positionInPlainBlock == Ciphers.PLAIN_BLOCK_SIZE) {
        encryptAndWriteBlock();
```

Reply:
This is really just to sanity check the state and avoid errors. The code is correct today because your assumption holds, but as this class changes, we want to make as many assumptions as possible explicit rather than relying on people knowing what to do.

Reply:
In my previous comment, I forgot about another call of this function - from the stream `close()` method. There, the last block is not assumed to be full, since this writes the remains of the plainBlock buffer.

Reply:
I've added a check making sure the block is not empty

Reply:
Actually, I meant that it would be good to have a check that only one partial block (last block) is written.